### PR TITLE
Upgrade React Router to V6

### LINKED ISSUE #805: Upgrade to react-router 6
We are using react-router 5 in our HTML table, and [version 6 has been released](https://remix.run/blog/react-router-v6). Upgrading needs some work on our side ([guide](https://reactrouter.com/docs/en/v6/upgrading/v5)).

We have to replace `<Switch>` with `<Routes>` and change a few things inside it, but this seems easy.

More complicated is that we use the `useHistory` hook to `listen` on history changes in `ReactTable.js`, and `useHistory` is gone now. We might be able to use [`useLocation`](https://reactrouter.com/docs/en/v6/api#uselocation) or the [`history` package](https://www.npmjs.com/package/history) that is used by `react-router`.

However, we also directly access the browser history in some other components and do `window.history.pushState()` and `window.addEventListener("popstate", ...)`, and sometimes we also access the `history` property from `react-router`, which also seems to be gone. So this also needs changes and we should probably rewrite all this history handling to use the same kind of code based on react-router.

### PR DESCRIPTION
**Pull request by EshaanAgg**
Reimplementation of all the code and discussion in #1054, which was closed due to wrong history errors. 

Fixes #805.

### CODE REVIEW

File: benchexec/tablegenerator/react-table/src/components/Overview.js [hunk 1]
```diff
@@ -370,62 +365,78 @@ export default class Overview extends React.Component {
...
-                  />
-                </Route>
-                <Route path="/info">
-                  <Info
-                    version={this.props.data.version}
-                    selectColumn={this.toggleSelectColumns}
-                  />
-                </Route>
-              </Switch>
+              {/* Component to sync the navigation state. Does not render anything. */}
```
Comment (PhilippWendler) [inline benchexec/tablegenerator/react-table/src/components/Overview.js:368]:
```suggestion
```

**Review by PhilippWendler — COMMENTED** [review]
Filters seem to work nicely.

However, when I try to change anything about a plot configuration, the URL updates but the rendered UI does not. I need to refresh the page to make it work. (Again in Firefox)

Can you reproduce this?

**Comment by EshaanAgg** [comment]
@PhilippWendler Thanks for the catch. I have pushed some changes, and now all the plots seem to be working on my end. Can you confirm the same?

**Review by PhilippWendler — APPROVED** [review]

**Comment by PhilippWendler** [comment]
Ah, now I found some more rendering problems: If I use the filters in the table header, it seems to render twice. For example, when I use the drop down, I first see a glimpse of the previous filter before it renders again. And if I use one of the filters with text entry, the rendering causes the text field to loose focus, so I cannot type continuously. Can you please have a look at this?

**Comment by EshaanAgg** [comment]
Can you please attach a video of the same? I tried to apply a `status` filter and then type a query in the `task name` field in the task filter, and it seems to work fine on my end on Firefox. Or am I not replicating it properly?

**Comment by PhilippWendler** [comment]
Sure:

https://github.com/sosy-lab/benchexec/assets/2545335/24f1e37f-7c2c-482d-9227-ad580cc1a823

This is on fee40a6f.

**Comment by EshaanAgg** [comment]
Got it, sir! I am trying the fixes for these and will update here as soon as I make progress.

**Comment by EshaanAgg** [comment]
I tried multiple approaches over the weekend, including:
- Trying to move the filter components to be top-level components and using their memorized versions to prevent rerenders
- Using a state variable called `focusedElement` to check the last focused element, and trying to restore the focus. 
- Changing the `useCallback` hooks used in multiple ways to prevent the re-renders
but I couldn't develop a solution that could preserve the focus without leading to some weird, incorrect filtering-related bugs. I am still trying my best here, but I am running out of ideas of how the same can be tackled. 

PS. I sincerely apologize for the extra time this PR is taking. The migration had posed various changes, which I had not initially envisioned even remotely.

**Comment by PhilippWendler** [comment]
Don't worry, and I think that you have done great work so far! I am sorry that I cannot help you with these technical problems. Unfortunately, the focus loss is kind of a big UI disadvantage, so we need to treat it as a blocker for merging it.

I remember that on the original implementation we also had some trouble to get the focus right, but eventually managed. Maybe you can find the relevant commits in the git history?

Do you think this creates a time problem for your project roadmap?

**Comment by EshaanAgg** [comment]
Yes sir. I am trying some other ways that could fix the issue. Thanks for all the support. I'll try to browse the history and what pops up. I don't think this should cause a problem for the roadmap right now, and I can make up for these weeks in the buffer time planned.

**Comment by EshaanAgg** [comment]
@PhilippWendler Can you please take a look? Fixed the focus problem by defining a state on the `ReactTable` component level and restoring it between re-renders. It also ensures that the same does not cause unnecessary re-renders by memoizing the component at each level.

**Comment by PhilippWendler** [comment]
Nice, thank you very much! I guess it is anyway better to have the `FilterInputFieldComponent`s as independent top-level definitions. Maybe we move them even into a separate file?

The only thing that is not perfect yet is the double re-rendering that happens when the drop-down is changed. Is this easy to fix? Otherwise I would say this is ready to merge. :tada:

**Comment by EshaanAgg** [comment]
Yes sir. Moving them to separate files is a better idea as it reduces the associated complexity of the base `ReactTable` component. I have done the same, as well as memorized the `StatusFilter` as well to avoid re-renders. Can you please confirm if the issue is now resolved (I can't seem to reproduce the re-rendering bug locally on either Chrome or Firefox, and thus I'm uncertain if it's fixed).

**Review by PhilippWendler — APPROVED** [review]
Very nice! Thank you a lot, this works really well! Congratulations for your good work! :rocket: 

I am glad that we can finally close this issue and solve this part of our technical debt backlog!

**Pull request merged**